plugins/backstage-plugin-devcontainers-backend/src/processors/DevcontainersProcessor.ts

```diff
@@ -0,0 +1,187 @@
+} from '@backstage/plugin-catalog-node';
+import { type Entity } from '@backstage/catalog-model';
+import { type Config } from '@backstage/config';
+import { isError, NotFoundError } from '@backstage/errors';
+import { type UrlReader, UrlReaders } from '@backstage/backend-common';
+import { type Logger } from 'winston';
+
+export const DEFAULT_TAG_NAME = 'devcontainers';
+export const PROCESSOR_NAME_PREFIX = 'backstage-plugin-devcontainers-backend/';
+
+type ProcessorOptions = Readonly<{
+  tagName: string;
+  logger: Logger;
+}>;
+
+type FromConfigOptions = Readonly<
+  Partial<ProcessorOptions> & {
+    logger: Logger;
+  }
+>;
+
+export class DevcontainersProcessor implements CatalogProcessor {
+  private readonly urlReader: UrlReader;
+  private readonly options: ProcessorOptions;
+
+  constructor(urlReader: UrlReader, options: ProcessorOptions) {
+    this.urlReader = urlReader;
+    this.options = options;
+  }
+
+  static fileLocations: readonly string[] = [
+    '.devcontainer/devcontainer.json',
+    '.devcontainer.json',
+  ];
+
+  static fromConfig(readerConfig: Config, options: FromConfigOptions) {
+    const processorOptions: ProcessorOptions = {
+      tagName: options.tagName || DEFAULT_TAG_NAME,
+      logger: options.logger,
+    };
+
+    const reader = UrlReaders.default({
+      config: readerConfig,
+      logger: options.logger,
+    });
+
+    return new DevcontainersProcessor(reader, processorOptions);
+  }
+
+  getProcessorName(): string {
+    // Very specific name to avoid name conflicts
+    return `${PROCESSOR_NAME_PREFIX}/devcontainers-processor`;
+  }
+
+  async preProcessEntity(
+    entity: Entity,
+    location: LocationSpec,
+    emit: CatalogProcessorEmit,
+  ): Promise<Entity> {
+    // The location of a component should be the catalog-info.yaml file, but
+    // check just to be sure.
+    const shouldNotProcess =
+      entity.kind !== 'Component' ||
+      !location.target.endsWith('/catalog-info.yaml');
+
+    if (shouldNotProcess) {
+      return entity;
+    }
+
+    // The catalog-info.yaml is not necessarily at the root of the repository.
+    // For showing the tag, we only care that there is a devcontainer.json
+    // somewhere in the catalog-info.yaml directory or below.  However, if this
+    // is a subdirectory (for example a monorepo) or a branch other than the
+    // default, VS Code will fail to open it.  We may need to skip adding the
+    // tag for anything that is not the root of the default branch, if we can
+    // get this information, or figure out a workaround.
+    const rootUrl = location.target.replace(/\/catalog-info\.yaml$/, '');
+
+    const entityLogger = this.options.logger.child({
+      name: entity.metadata.name,
+      rootUrl,
+    });
+
+    try {
+      const jsonUrl = await this.findDevcontainerJson(rootUrl, entityLogger);
+      entityLogger.info('Found devcontainer config', { url: jsonUrl });
+      return this.addTag(entity, this.options.tagName, entityLogger);
```

Comment:
Main change here was making sure we weren't always passing in `DEFAULT_TAG_NAME` to this method call

Reply:
Good catch!!